temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+                self._heartbeat_activity_async(
+                    logger, task_token, heartbeat_seq, *details
+                )
+            )
+
+    async def _heartbeat_activity_async(
+        self,
+        logger: logging.Logger,
+        task_token: bytes,
+        heartbeat_seq: int,
+        *details: Any,
+    ) -> None:
+        activity = self._running_activities.get(task_token)
+        # Bail if not the latest
+        if not activity or activity.current_heartbeat_seq != heartbeat_seq:
+            return
+        try:
+            heartbeat = temporalio.bridge.proto.ActivityHeartbeat(task_token=task_token)
+            if details:
+                heartbeat.details.extend(
+                    temporalio.bridge.worker.payloads_to_core(
+                        await self._config["client"].data_converter.encode(details)
+                    )
+                )
+            # Bail if not the latest
+            if activity.current_heartbeat_seq != heartbeat_seq:
+                return
+            self._bridge_worker.record_activity_heartbeat(heartbeat)
+        except Exception as err:
+            # Bail if not the latest
+            if activity.current_heartbeat_seq != heartbeat_seq:
+                return
+            # If the activity is done, nothing we can do but log
+            if activity.done:
+                logger.exception(
+                    f"Failed recording heartbeat (activity already done, cannot error): {err}",
+                    extra={"error": err},
+                )
+            else:
+                logger.warning(
+                    f"Cancelling activity because failed recording heartbeat: {err}",
+                    extra={"error": err},
+                    exc_info=True,
+                )
+                activity.cancel(cancelled_due_to_heartbeat_error=err)
+
+    async def _run_activity(
+        self,
+        task_token: bytes,
+        start: temporalio.bridge.proto.activity_task.Start,
+        running_activity: _RunningActivity,
+    ) -> None:
+        logger.debug("Running activity %s (token %s)", start.activity_type, task_token)
+        # We choose to surround interceptor creation and activity invocation in
+        # a try block so we can mark the workflow as failed on any error instead
+        # of having error handling in the interceptor
+        completion = temporalio.bridge.proto.ActivityTaskCompletion(
+            task_token=task_token
+        )
+        try:
+            # Find activity or fail
+            activity_def = self._activities.get(start.activity_type)
+            if not activity_def:
+                activity_names = ", ".join(sorted(self._config["activities"].keys()))
+                raise temporalio.exceptions.ApplicationError(
+                    f"Activity function {start.activity_type} is not registered on this worker, available activities: {activity_names}",
+                    type="NotFoundError",
+                    non_retryable=True,
```

Comment:
I don't remember what we decided here. should this be non retryable?

Reply:
I took this from https://github.com/temporalio/sdk-typescript/blob/v0.18.0/packages/worker/src/worker.ts#L474. If it is not right there and needs to be fixed there, I can incorporate.

Reply:
Please check all of these non retryable errors in other SDKs, I'm not sure TS is doing it correctly.
If we're doing it wrong in TS lets open an issue.

Reply:
A bit of a double standard on pre-merge SDK quality here. One wonders if the entire TS codebase should be re-reviewed for things that are being preemptively reviewed in this SDK?

I will investigate other SDKs and get back.

Reply:
After investigation, opened https://github.com/temporalio/sdk-typescript/issues/521.

Reply:
Changed